frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseResetBranchToRemoteAction.java

```diff
@@ -0,0 +1,179 @@
+    if (branchName.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled due to undefined ${getActionType()} branch");
+      return;
+    }
+
+    String branchNameString = branchName.get();
+
+    Option<SyncToRemoteStatus> syncToRemoteStatus = getGitMacheteRepository(anActionEvent)
+        .flatMap(repo -> repo.getBranchByName(branchNameString))
+        .map(BaseGitMacheteBranch::getSyncToRemoteStatus);
+
+    if (syncToRemoteStatus.isEmpty()) {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled due to undefined sync to remote status");
+      return;
+    }
+
+    SyncToRemoteStatus.Relation relation = syncToRemoteStatus.get().getRelation();
+
+    if (relation != SyncToRemoteStatus.Relation.Untracked) {
+      presentation.setDescription("Reset branch '${branchNameString} to its remote tracking branch");
+    } else {
+      presentation.setEnabled(false);
+      presentation.setDescription("Reset disabled because ${getActionType()} branch is untracked");
+    }
+  }
+
+  @Override
+  @UIEffect
+  public void actionPerformed(AnActionEvent anActionEvent) {
+    LOG.debug("Performing");
+
+    Project project = getProject(anActionEvent);
+    var gitRepository = getSelectedVcsRepository(anActionEvent);
+    var branchName = getNameOfBranchUnderAction(anActionEvent);
+    var macheteRepository = getGitMacheteRepository(anActionEvent);
+
+    if (branchName.isDefined()) {
+      if (gitRepository.isDefined()) {
+        if (macheteRepository.isDefined()) {
+          doResetToRemoteWithKeep(project, gitRepository.get(), branchName.get(), macheteRepository.get(),
+              anActionEvent);
+        } else {
+          LOG.error("Skipping the action because can't get Git Machete repository");
+          VcsNotifier.getInstance(project).notifyError(VCS_NOTIFIER_TITLE,
+              "Internal error occurred. For more information see IDE log file");
+        }
+      } else {
+        LOG.warn("Skipping the action because no VCS repository is selected");
+        VcsNotifier.getInstance(project).notifyWarning(VCS_NOTIFIER_TITLE,
+            "Skipping the action because no VCS repository is selected");
+      }
+    } else {
+      LOG.warn("Skipping the action because name of branch to reset is undefined");
+      VcsNotifier.getInstance(project).notifyError(VCS_NOTIFIER_TITLE,
+          "Internal error occurred. For more information see IDE log file");
+    }
+  }
+
+  protected void doResetToRemoteWithKeep(Project project, GitRepository gitRepository, String branchName,
+      IGitMacheteRepository macheteRepository, AnActionEvent anActionEvent) {
+
+    new Task.Backgroundable(project, TASK_TITLE, /* canBeCancelled */ true) {
+
+      @Override
+      public void run(ProgressIndicator indicator) {
+        LOG.debug(() -> "Resetting '${branchName}' branch");
+        try (AccessToken ignored = DvcsUtil.workingTreeChangeStarted(project, TASK_TITLE)) {
```

Comment:
Drop a comment on why this is necessary

Reply:
TBH, I'm not 100% but from JB code it seems to be try with resources construction that notify listeners about batch changes and probably it's some kind of `mutex` for repo operations (see: https://github.com/JetBrains/intellij-community/blob/master/platform/dvcs-impl/src/com/intellij/dvcs/DvcsUtil.java#L171)
This piece of code is also JB actions, so I think that it can be important.